File: backend/LazyRisparmioforo/LazyRisparmioforo.Services/TransactionService/TransactionService.cs
using LazyRisparmioforo.Domain.Commands;
using LazyRisparmioforo.Domain.Entities;
using LazyRisparmioforo.Domain.Queries;
using LazyRisparmioforo.Infrastructure.Data;
using LazyRisparmioforo.Shared.Shared;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Result = LazyRisparmioforo.Shared.Shared.Result;

namespace TransactionService;

public class TransactionService(
    ApplicationDbContext dbContext,
    ILogger<TransactionService> logger)
    : ITransactionService
{
    public async Task<Pagination<Transaction>> SearchAsync(TransactionSearchQuery command, CancellationToken cancellationToken)
    {
        var query = dbContext.Transactions
            .AsNoTracking()
            .Where(t =>
                (command.Flow == null || t.Flow == command.Flow) && 
                (command.FromDate == null || t.RegistrationDate >= command.FromDate) &&
                (command.ToDate == null || t.RegistrationDate <= command.ToDate) &&
                (string.IsNullOrEmpty(command.Query) || t.Description.ToLower().Contains(command.Query.ToLower())))
            .OrderByDescending(x => x.RegistrationDate)
            .AsQueryable();

        var totalItemsCount = await query.CountAsync(cancellationToken);

        var items = await query
            .Skip(command.PageSize * command.PageIndex)
            .Take(command.PageSize)
            .ToListAsync(cancellationToken);

        // var dto = items.ToDto();
        var paginatedResults = items.ToPagination(command.PageIndex, command.PageSize, totalItemsCount);

        return paginatedResults;
    }

    public async Task<Result<Transaction>> GetAsync(TransactionGetCommand command, CancellationToken cancellationToken)
    {
        var entity = await dbContext.Transactions
            .AsNoTracking()
            .FirstOrDefaultAsync(transaction => transaction.Id == command.Id, cancellationToken);

        return entity is null
            ? Result.Failure<Transaction>(Errors.NotFound(command.Id))
            : Result.Success(entity);
    }

    public async Task<Result> CreateAsync(TransactionCreateCommand command, CancellationToken cancellationToken)
    {
        var entity = command.ToEntity();

        dbContext.Transactions.Add(entity);
        await dbContext.SaveChangesAsync(cancellationToken);

        logger.LogInformation("Transaction created successfully.");

        return Result.Success();
    }

    public async Task<Result> UpdateAsync(TransactionUpdateCommand command, CancellationToken cancellationToken)
    {
        var entity = await dbContext.Transactions
            .AsNoTracking()
            .FirstOrDefaultAsync(x => x.Id == command.Id, cancellationToken);

        if (entity is null)
            return Result.Failure(Errors.NotFound(command.Id));

        entity = command.ToEntity(entity);

        dbContext.Entry(entity).State = EntityState.Modified;
        dbContext.Transactions.Update(entity);
        await dbContext.SaveChangesAsync(cancellationToken);

        logger.LogInformation("Transaction updated successfully.");

        return Result.Success();
    }

    public async Task<Result> RemoveAsync(TransactionRemoveCommand command, CancellationToken cancellationToken)
    {
        var entity = await dbContext.Transactions
            .FirstOrDefaultAsync(x => x.Id == command.Id, cancellationToken);

        if (entity is null)
            return Result.Failure(Errors.NotFound(command.Id));

        dbContext.Transactions.Remove(entity);

        await dbContext.SaveChangesAsync(cancellationToken);

        logger.LogInformation("Transaction removed successfully.");

        return Result.Success();
    }
}

File: frontend/LazyRisparmioforo.UI/src/app/services/transaction-service/transaction.service.ts
import {inject, Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {environment} from '../../../environments/environment';
import {Observable} from 'rxjs';
import {
  Transaction,
  TransactionGetCommand,
  TransactionSearchCommand,
  TransactionCreateCommand,
  TransactionUpdateCommand,
  TransactionRemoveCommand,
} from './transaction.model';
import {Pagination} from '../../shared/models/pagination';

@Injectable({
  providedIn: 'root'
})
export class TransactionService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/transactions`;

  get(request: TransactionGetCommand): Observable<Transaction> {
    return this.http.get<Transaction>(`${this.apiUrl}/${request.id}`);
  }

  search(request: TransactionSearchCommand): Observable<Pagination<Transaction>> {
    let url = `${this.apiUrl}/search?PageIndex=${request.pageIndex}&PageSize=${request.pageSize}`;
    if (request.flow != undefined) url = url + `&Flow=${request.flow}`;
    if (request.query != undefined) url = url + `&Query=${request.query}`;
    if (request.fromDate != undefined) url = url + `&FromDate=${request.fromDate}`;
    if (request.toDate != undefined) url = url + `&ToDate=${request.toDate}`;
    return this.http.get<Pagination<Transaction>>(url);
  }

  create(request: TransactionCreateCommand): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/create`, request);
  }

  update(request: TransactionUpdateCommand): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/update`, request);
  }

  delete(request: TransactionRemoveCommand): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/remove?Id=${request.id}`);
  }
}
